### PR TITLE
Fix ServiceInstance selection with labels

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
This is a regression recently introduced.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -1921,16 +1921,30 @@ func (ps *PushContext) BestEffortInferServiceMTLSMode(tp *networking.TrafficPoli
...
 		if instances, exists := ps.ServiceIndex.instancesByPort[svc][port]; exists {
 			return instances
 		}
 	}
-	return nil
+
+	out := []*ServiceInstance{}
+	if instances, exists := ps.ServiceIndex.instancesByPort[svc][port]; exists {
+		// Use cached version of instances by port when labels are empty.
+		if len(labels) == 0 {
```
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go:1935]: Is this still needed here?

File: pilot/pkg/model/push_context.go [hunk 2]
```diff
@@ -1921,16 +1921,30 @@ func (ps *PushContext) BestEffortInferServiceMTLSMode(tp *networking.TrafficPoli
...
-	// we will have to make actual call and filter instances by pod labels.
 	if len(labels) == 0 {
 		if instances, exists := ps.ServiceIndex.instancesByPort[svc][port]; exists {
 			return instances
 		}
 	}
-	return nil
+
+	out := []*ServiceInstance{}
+	if instances, exists := ps.ServiceIndex.instancesByPort[svc][port]; exists {
```
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go:1933]: Can we get this above? and use it here also?

**Comment by howardjohn** [comment]
> This is a regression recently introduced.

By which PR?

**Review by howardjohn — APPROVED** [review]
LGTM, what versions is this broken on?

**Comment by hzxuzhonghu** [comment]
broken in 1.11

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:1935]: good

**Comment by hzxuzhonghu** [comment]
@ramaraochavali addressed

**Review by ramaraochavali — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
Thanks, i removed the hold

**Pull request merged**